File: Top_100_Liked_Questions/17._Letter_Combinations_of_a_Phone_Number/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        
        d2c = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        
        ans = [""]
        
        for digit in digits:
            chars = d2c[digit]
            newAns = []
            for item in ans:
                newAns += [item] * len(chars)
            ans = newAns
            
            for i in range(len(ans)):
                ans[i] += chars[i%len(chars)]
        
        return ans

File: Top_100_Liked_Questions/17._Letter_Combinations_of_a_Phone_Number/solution.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0)
            return new ArrayList<>();

        Map<Character, List<Character>> num2char = new HashMap<>();
        num2char.put('2', new ArrayList<Character>(List.of('a', 'b', 'c')));
        num2char.put('3', new ArrayList<Character>(List.of('d', 'e', 'f')));
        num2char.put('4', new ArrayList<Character>(List.of('g', 'h', 'i')));
        num2char.put('5', new ArrayList<Character>(List.of('j', 'k', 'l')));
        num2char.put('6', new ArrayList<Character>(List.of('m', 'n', 'o')));
        num2char.put('7', new ArrayList<Character>(List.of('p', 'q', 'r', 's')));
        num2char.put('8', new ArrayList<Character>(List.of('t', 'u', 'v')));
        num2char.put('9', new ArrayList<Character>(List.of('w', 'x', 'y', 'z')));

        List<String> ans = new ArrayList<>();
        ans.add("");
        for (int i=0; i<digits.length(); i++) {
            List<Character> chars = num2char.get(digits.charAt(i));
            List<String> currAns = new ArrayList<>(ans);
            ans.clear();
            for (Character ch : chars) {
                for (String s : currAns) {
                    ans.add(s + ch);
                }
            }
        }

        return ans;
    }
}